File: Algorithms/131. Palindrome Partitioning.rs
impl Solution {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        let mut result: Vec<Vec<String>> = vec![];
        let s: Vec<char> = s.chars().collect();
        let length = s.len();
        let mut indexes: Vec<(usize, usize)> = vec![];
        Self::dfs(0, &mut indexes, &mut result, &s, length);
        result
    }
    
    fn dfs(start: usize, indexes: &mut Vec<(usize, usize)>, result: &mut Vec<Vec<String>>, s: &[char], length: usize) {
        if start == length {
            let mut partition: Vec<String> = vec![];
            for &(l, r) in indexes.iter() {
                partition.push(s[l..r].iter().collect());
            }
            result.push(partition);
        }
        
        for end in start + 1..=length {
            if Self::is_palindrome(&s[start..end]) {
                indexes.push((start, end));
                Self::dfs(end, indexes, result, s, length);
                indexes.pop();
            }
        }
    }
    
    fn is_palindrome(s: &[char]) -> bool {
        let n = s.len();
        for i in 0..n {
            let j = n - 1 - i;
            if s[i] != s[j] {
                return false;
            }
        }
        true
    }
}

File: Algorithms/131. Palindrome Partitioning.ts
function partition(s: string): string[][] {
  const n = s.length;
  const dp: boolean[][] = Array.from({ length: n }, (_) =>
    new Array(n).fill(true)
  );
  for (let i = n - 2; i >= 0; i--) {
    for (let j = i + 1; j < n; j++) {
      dp[i][j] = s[i] === s[j] && dp[i + 1][j - 1];
    }
  }
  const result: string[][] = [];
  const temp: string[] = [];
  const dfs = (i: number) => {
    if (i === n) {
      result.push(temp.slice());
      return;
    }
    for (let j = i; j < n; j++) {
      if (dp[i][j]) {
        temp.push(s.slice(i, j + 1));
        dfs(j + 1);
        temp.pop();
      }
    }
  };
  dfs(0);
  return result;
}
